zds/notification/models.py

```diff
@@ -26,25 +23,274 @@ class Meta:
+            notification = Notification(subscription=self, content_object=content, sender=sender)
+            notification.url = self.get_notification_url(content)
+            notification.title = self.get_notification_title(content)
+            notification.save()
+            self.set_last_notification(notification)
+            self.save()
+
+            if send_email & self.by_email:
+                self.send_email(notification)
+        elif self.last_notification is not None:
+            # Update last notif if the new content is older (case of unreading)
+            if not self.last_notification.is_read and self.last_notification.pubdate > content.pubdate:
+                self.last_notification.content_object = content
+                self.last_notification.save()
+
+    def mark_notification_read(self):
+        if self.last_notification is not None:
+            self.last_notification.is_read = True
+            self.last_notification.save()
+
+
+class MultipleNotificationsMixin(object):
+
+    def send_notification(self, content=None, send_email=True, sender=None):
+        assert hasattr(self, "get_notification_url")
+        assert hasattr(self, "get_notification_title")
+        assert hasattr(self, "send_email")
+
+        notification = Notification(subscription=self, content_object=content, sender=sender)
+        notification.url = self.get_notification_url(content)
+        notification.title = self.get_notification_title(content)
+        notification.save()
+        self.set_last_notification(notification)
+
+        if send_email & self.by_email:
+            self.send_email(notification)
+
+    def mark_notification_read(self, content):
+        if content is None:
+            raise Exception('Object content of notification must be defined')
+
+        content_notification_type = ContentType.objects.get_for_model(content)
+        try:
+            notification = Notification.objects.get(subscription=self,
+                                                    content_type__pk=content_notification_type.pk,
+                                                    object_id=content.pk, is_read=False)
+        except Notification.DoesNotExist:
+            notification = None
+        if notification is not None:
+            notification.is_read = True
+            notification.save()
+
+
+class AnswerSubscription(Subscription):
+    """
+    Subscription to new answer, either in a topic, a article or a tutorial
+    """
+
+    def __unicode__(self):
+        return _(u'<Abonnement du membre "{0}" aux réponses au {1} #{2}>')\
+            .format(self.profile, self.content_type, self.object_id)
+
+    def get_notification_url(self, answer):
+        return answer.get_absolute_url()
+
+    def get_notification_title(self, answer):
+        return self.content_object.title
+
+
+
+class TopicAnswerSubscription(AnswerSubscription, SingleNotificationMixin):
+
+    module = _(u'Forum')
+    objects = SubscriptionManager()
```

Comment:
C'est faisable, mais ça implique de déclarer la classe de base comme abstraite. https://docs.djangoproject.com/en/dev/topics/db/managers/#custom-managers-and-model-inheritance

Du coup, ça limite ce qu'il est possible de faire avec la classe Subscription. Par exemple, on ne peut plus l'avoir dans l'interface admin. Ca implique aussi de chipoter avec les attributs clés étrangères : https://docs.djangoproject.com/fr/1.8/topics/db/models/#abstract-base-classes

L'un dans l'autre, ça me semble mieux de leur filer le manager à toutes. Mais je ne suis pas encore très à l'aise avec la notion d'héritage dans python donc je dis peut-être des bêtises.